erpnext/accounts/report/gross_and_net_profit_report/gross_and_net_profit_report.py

```diff
@@ -0,0 +1,166 @@
+	elif revenue_type == 'non_gross':
+		revenue = [item for item in data if item['include_in_gross']==0 or item['is_group']==1]
+	revenue, status = remove_parent_with_no_child(revenue, period_list)
+	while status == "data to be removed":
+		revenue, status = remove_parent_with_no_child(revenue, period_list)
+	revenue = adjust_account(revenue, period_list)
+	return copy.deepcopy(revenue)
+
+def remove_parent_with_no_child(data, period_list):
+	status = "nothing to remove"
+	for parent in data:
+		if 'is_group' in parent and parent["is_group"] == 1:
+			have_child = False
+			for child in data:
+				if 'parent_account' in child  and child["parent_account"] == parent["account"]:
+					have_child = True
+					break
+
+			if not have_child:
+				status = "data to be removed"
+				data.remove(parent)
+
+	return data, status
+
+def adjust_account(data, period_list, consolidated= False):
+	leaf_nodes = [item for item in data if item['is_group'] == 0]
+	totals = {}
+	for node in leaf_nodes:
+		set_total(node, node["total"], data, totals)
+	for d in data:
+		for period in period_list:
+			key = period if consolidated else period.key
+			d[key] = totals[d["account"]]
+			d['total'] = totals[d["account"]]
+	return data
+
+def set_total(node, value, complete_list, totals):
+	if not totals.get(node['account']):
+		totals[node["account"]] = 0
+	totals[node["account"]] += value
+
+	parent = node['parent_account']
+	if not parent == '':
+		return set_total(next(item for item in complete_list if item['account'] == parent), value, complete_list, totals)
+
+
+def get_profit(gross_income, gross_expense, period_list, company, profit_type, currency=None, consolidated=False):
+
+	total = 0
+
+	profit_loss = {
+		"account_name": "'" + _(profit_type) + "'",
+		"account": "'" + _(profit_type) + "'",
+		"warn_if_negative": True,
+		"currency": currency or frappe.get_cached_value('Company',  company,  "default_currency")
+	}
+
+	has_value = False
+
+	for period in period_list:
+		key = period if consolidated else period.key
+		profit_loss[key] = flt(gross_income[0][key] if len(gross_income) else 0) - flt(gross_expense[0][key] if len(gross_expense) else 0)
+
+		if profit_loss[key]:
+			has_value=True
+
+		total += flt(profit_loss[key])
+		profit_loss['total'] = total
+
+	if has_value:
+		return profit_loss
+
+def get_net_profit(non_gross_income, gross_income, gross_expense, non_gross_expense, period_list, company, currency=None, consolidated=False):
+	total = 0
+	profit_loss = {
+		"account_name": "'" + _("Net Profit") + "'",
+		"account": "'" + _("Net Profit") + "'",
+		"warn_if_negative": True,
+		"currency": currency or frappe.get_cached_value('Company',  company,  "default_currency")
+	}
+
+	has_value = False
+
+	for period in period_list:
+		key = period if consolidated else period.key
+		total_income = flt(gross_income[0][key] if len(gross_income) else 0) + flt(non_gross_income[0][key] if len(non_gross_income) else 0)
```

Comment:
gross_income[0][key] not looks good
From method get_revenue, get the value directly instead of list and pass the same value to the get_net_profit so you don't have to check if len(gross_income)

Reply:
done